prow.sh

```diff
@@ -0,0 +1,916 @@
+# then `make WHAT=test/e2e/e2e.test` is called first to ensure that
+# all generated files are present.
+#
+# CSI_PROW_E2E_REPO=none disables E2E testing.
+configvar CSI_PROW_E2E_VERSION v1.14.0 "E2E version"
+configvar CSI_PROW_E2E_REPO https://github.com/kubernetes/kubernetes "E2E repo"
+configvar CSI_PROW_E2E_IMPORT_PATH k8s.io/kubernetes "E2E package"
+
+# csi-sanity testing from the csi-test repo can be run against the installed
+# CSI driver. For this to work, deploying the driver must expose the Unix domain
+# csi.sock as a TCP service for use by the csi-sanity command, which runs outside
+# of the cluster. The alternative would have been to (cross-)compile csi-sanity
+# and install it inside the cluster, which is not necessarily easier.
+configvar CSI_PROW_SANITY_REPO https://github.com/kubernetes-csi/csi-test "csi-test repo"
+configvar CSI_PROW_SANITY_VERSION 5421d9f3c37be3b95b241b44a094a3db11bee789 "csi-test version" # latest master
+configvar CSI_PROW_SANITY_IMPORT_PATH github.com/kubernetes-csi/csi-test "csi-test package"
+configvar CSI_PROW_SANITY_SERVICE "hostpath-service" "Kubernetes TCP service name that exposes csi.sock"
+configvar CSI_PROW_SANITY_POD "csi-hostpathplugin-0" "Kubernetes pod with CSI driver"
+configvar CSI_PROW_SANITY_CONTAINER "hostpath" "Kubernetes container with CSI driver"
+
+# Each job can run one or more of the following tests, identified by
+# a single word:
+# - unit testing
+# - parallel excluding alpha features
+# - serial excluding alpha features
+# - parallel, only alpha feature
+# - serial, only alpha features
+# - sanity
+#
+# Sanity testing with csi-sanity only covers the CSI driver itself and thus
+# is off by default. A CSI driver can change that default in its .prow.sh
+# by setting CSI_PROW_TESTS_SANITY.
+configvar CSI_PROW_TESTS "unit parallel serial parallel-alpha serial-alpha ${CSI_PROW_TESTS_SANITY}" "tests to run"
+test_enabled () {
+    echo "${CSI_PROW_TESTS}" | grep -q -w -e "$1"
+}
+
+# Serial vs. parallel is always determined by these regular expressions.
+# Individual regular expressions are seperated by spaces for readability
+# and expected to not contain spaces. Use dots instead. The complete
+# regex for Ginkgo will be created by joining the individual terms.
+configvar CSI_PROW_E2E_SERIAL '\[Serial\] \[Disruptive\]' "tags for serial E2E tests"
+regex_join () {
+    echo "$@" | sed -e 's/  */|/g' -e 's/^|*//' -e 's/|*$//' -e 's/^$/this-matches-nothing/g'
+}
+
+# Prints the value of a variable + version suffix, falling back to variable + "LATEST".
+get_versioned_variable () {
+    local var="$1"
+    local version="$2"
+    local value
+
+    eval value="\${${var}_${version}}"
+    if ! [ "$value" ]; then
+        eval value="\${${var}_LATEST}"
+    fi
+    echo "$value"
+}
+
+# Which tests are alpha depends on the Kubernetes version. The same
+# E2E test suite is used for all Kubernetes versions, including older
```

Comment:
We should not make this assumption.  This is going to break very quickly any time we move a feature from alpha to beta, and wastes many hours of developer time trying to debug what's wrong.  

For beta and GA features, the e2e test version that is used must be <= the K8s version.
For alpha features, the e2e test version must == K8s version

We may need to special case 1.13 because the test framework was not completed until 1.14, but this should not be the norm.  The other option is to leave 1.13->1.14 version skew testing to in-tree k/k, and just start the kubernetes-csi CI starting from 1.14.

Reply:
I don't agree, but so be it. It's not hard to make the `CSI_PROW_E2E` settings depend on the Kubernetes version, I just don't have time for that right now. Can we defer this?

Reply:
if we don't have time to do this, then I prefer we leave out skew testing for now, and have 1.14 against 1.14, and master against master, where the e2e test version is the same as the kubernetes version. Let's take incremental steps instead of trying to tackle every single scenario at once.

The way that is right now is going to require considerable effort next release to update the versions used in the tests.

Reply:
I'll do it tomorrow. We are almost there, I don't want to go back.